Initial implementation

mcdonnnj opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This is essentially an implementation of [cisagov/cool-users-non-admin](https://github.com/cisagov/cool-users-non-admin) but targeting the legacy CyHy environment. The self-management policies are lifted directly from the `users` account configuration in [cisagov/cool-accounts](https://github.com/cisagov/cool-accounts).
<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

Current user creation is spread out across multiple projects ([cisagov/findings-data-s3-users](https://github.com/cisagov/findings-data-s3-users) and [cisagov/bod-18-01-cloudwatch-users](https://github.com/cisagov/bod-18-01-cloudwatch-users)) and I am planning on making a new project to provide access to manage the users for the bastion instance. In addition to making it difficult to track non-admin users it makes it impossible for the same user to be in multiple permission groups. This has not been an issue in the past because of the low number of IAM accounts in the legacy CyHy environment, but it could be a lurking problem.
<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## ⚠ Implementation Notes ##

I will need to make changes to [cisagov/bod-18-01-cloudwatch-users](https://github.com/cisagov/bod-18-01-cloudwatch-users) and [cisagov/findings-data-s3-users](https://github.com/cisagov/findings-data-s3-users) to remove the user creation elements and consolidate all accounts on the user policies added in this repository.

## 🧪 Testing ##

Automated tests pass.
<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] All new and existing tests pass.

mcdonnnj committed:
Remove the examples/ directory tree

mcdonnnj committed:
Update base configuration

Add the appropriate base configuration for the CyHy environment.

mcdonnnj committed:
Add user self-management policies

Add the user self-management policies for both with and without MFA.
These are copied from:
https://github.com/cisagov/cool-accounts/tree/develop/users

mcdonnnj committed:
Add the user creation configuration

This configuration is sourced (with minor edits) from:
https://github.com/cisagov/cool-users-non-admin/blob/develop/users.tf

mcdonnnj committed:
Update the README

Update the README to reflect the changes made to this configuration.

mcdonnnj committed:
Enable Dependabot ignore directive

This enables the Dependabot ignore directive for the root directory.

jsf9k commented on variables.tf:
```diff
@@ -4,30 +4,55 @@
 }
 
-variable "aws_availability_zone" {
+variable "non_self_admin_users" {
+  default     = []
+  description = "A list containing the usernames of non-admin users that are not allowed to administer their own accounts.  Example: [ \"service-account1\", \"service-account2\", \"service-account3\" ]"
+  type        = list(string)
+}
```
Any thoughts on adding this capability as another attribute in the `object` associated with each `username` in `var.users`?  That seems cleaner and less confusing, although it moves the complication to the Terraform code.

mcdonnnj replied:
I would be ok with that if we wanted to go that route. Something like a `bool` attribute called `self_admin` or something along those lines? What do you think @dav3r?

dav3r replied:
Sure, that suggestion makes sense to me- I say go for it.

jsf9k replied:
Assuming it works out, we should make the same changes to [cisagov/cool-users-non-admin](https://github.com/cisagov/cool-users-non-admin).

mcdonnnj replied:
With the approval of https://github.com/cisagov/cool-users-non-admin/pull/34 I have made similar changes in [eff3553](https://github.com/cisagov/cyhy-users-non-admin/pull/5/commits/eff35534dc9004ca10d45f7de26f37ada34258eb).

jsf9k commented on variables.tf:
```diff
@@ -4,30 +4,55 @@
-variable "subnet_id" {
-  type        = string
-  description = "The ID of the AWS subnet to deploy into (e.g. subnet-0123456789abcdef0)."
+variable "users" {
+  description = "A map whose keys are the usernames of each non-admin user and whose values are a map containing supported user attributes.  The only currently-supported attribute is \"require_mfa\" (boolean).  Example: { \"firstname1.lastname1\" = { \"require_mfa\" = false }, \"firstname2.lastname2\" = { \"require_mfa\" = true }, \"firstname3.lastname3\" = { \"require_mfa\" = false } }"
+  type        = map(map(string))
```
This variable is currently of type `map(map(bool))` and would become more complicated if [this suggestion](https://github.com/cisagov/cyhy-users-non-admin/pull/5#discussion_r819646945) were implemented.  I think it makes sense to define it more precisely:
```suggestion
variable "users" {
  description = "A map whose keys are the usernames of each non-admin user and whose values are an object containing supported user attributes.  The only currently-supported attribute is \"require_mfa\" (boolean).  Example: { \"firstname1.lastname1\" = { \"require_mfa\" = false }, \"firstname2.lastname2\" = { \"require_mfa\" = true }, \"firstname3.lastname3\" = { \"require_mfa\" = false } }"
  type        = map(object({require_mfa = bool}))
```

mcdonnnj replied:
Please see [5c7c918](https://github.com/cisagov/cyhy-users-non-admin/pull/5/commits/5c7c918fd83aa08c8ca748cecbb16d997d0eecc2).

jsf9k reviewed (changes requested):
I have a requested change (which should be made to [cisagov/cool-users-non-admin](https://github.com/cisagov/cool-users-non-admin) as well) and a question.

mcdonnnj committed:
Consolidate variables that provider users

Remove the `non_self_admin_users` variable and integrate the same
functionality into the `users` variable by adding a new `self_managed`
attribute.

mcdonnnj committed:
Update the tfvars example in the README

Update the tfvars example that is in the README to mirror the example
given in the `users` variable description.

mcdonnnj committed:
Be more specific with the users variable type

Use an `map(object())` type with the specific attributes that are
supported instead of a blanket `map(map())` type.

jsf9k reviewed (approved):
Thanks for making those changes, both here and in the [cisagov/cool-users-non-admin](https://github.com/cisagov/cool-users-non-admin) repo!

dav3r reviewed (approved):
👍 👍  What @jsf9k said!  💪 💼
